sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/V1Writes.scala

```diff
@@ -107,7 +107,11 @@ object V1Writes extends Rule[LogicalPlan] with SQLConfHelper {
     if (orderingMatched) {
       empty2NullPlan
     } else {
-      Sort(requiredOrdering, global = false, empty2NullPlan)
+      // Preserve the requiredOrdering and append any missing elements from outputOrdering.
+      val newOrder = requiredOrdering ++ outputOrdering.filterNot { outputOrder =>
+        requiredOrdering.exists(_.child.semanticEquals(outputOrder.child))
```

Comment:
A confused case is: if people do `sort by col1, part1`, then only the `col1` will be added. However, the final data ordering is `part1, col1` rather than `col1, part1`, which still breaks people expect.

Reply:
It should be fine since we are writing data belonging to the same partition or bucket into the same files. In your example, the data within each file of a partition is sorted by `col1`, as expected.